src/Device.ts

```diff
@@ -58,21 +63,131 @@ class Main {
             {
                 onOff: false, // Off by default
                 globalSceneControl: true,
-                onTime: 0,
-                offWaitTime: 0,
-                startUpOnOff: OnOffStartUpOnOff.Off
+                onTime: 0, // unit: 1/10ths second
+                offWaitTime: 0, // unit: 1/10ths second
+                startUpOnOff: OnOffStartUpOnOff.Off // null?
             },
             {
-                on: async ({ attributes: { onOff } }) => onOff.set(true),
-                off: async ({ attributes: { onOff } }) => onOff.set(false),
-                toggle: async ({ attributes: { onOff } }) => onOff.set(!onOff.get()),
-                offWithEffect: async ({ attributes: { onOff } }) => Promise.resolve(), // TODO
-                onWithRecallGlobalScene: async ({ attributes: { onOff } }) => Promise.resolve(), // TODO
-                onWithTimedOff: async ({ attributes: { onOff } }) => Promise.resolve(), // TODO
+                on: async ({ attributes: { onOff, onTime, offWaitTime /*, globalSceneControl*/ } }) => {
+                    onOff.set(true);
+                    // globalSceneControl.set(true);
+                    if (onTime.get() === 0) {
+                        offWaitTime.set(0);
+                    }
+                },
+                off: async ({ attributes: { onOff, onTime } }) => {
+                    onOff.set(false);
+                    onTime.set(0);
+                },
+                toggle: async ({ attributes: { onOff, onTime, offWaitTime } }) => {
+                    if (onOff.get()) {
+                        onOff.set(false);
+                        onTime.set(0);
+                    } else {
+                        onOff.set(true);
+                        if (onTime.get() === 0) {
+                            offWaitTime.set(0);
+                        }
+                    }
+                },
+                offWithEffect: async ({request: { effectId, effectVariant}, attributes: { onOff/*, globalSceneControl*/ } }) => {
```

Comment:
Ut seems that because of "non Writable Attribute" definition for "globalSceneControl" I can not get it as attributes here - at least my IDE and TS say it is not allowed.
But "Writeable" for attributes just mean that they can not be written by a request from "external" ... but they still needs to be available internally and also writeable

Reply:
Works that way, thank you

Reply:
Just for info: UseOptionalAttributes worked when I used it on the Device.ts ... but when I started moving the methods in to a matter/cluster/server/OnOffServer.ts with a "OnOffClusterHandler" struct like theothers there then I did not found the place to use this above trick. how it would need to look like?